gymnasium/envs/tabular/blackjack.py

```diff
@@ -0,0 +1,477 @@
+"""This module provides a Blackjack functional environment and Gymnasium environment wrapper BlackJackJaxEnv."""
+
+
+import math
+import os
+from typing import Optional, Tuple, Union
+
+import jax
+import jax.numpy as jnp
+import numpy as np
+from jax import random
+from jax.random import PRNGKey
+
+from gymnasium import spaces
+from gymnasium.error import DependencyNotInstalled
+from gymnasium.experimental.functional import ActType, FuncEnv, StateType
+from gymnasium.experimental.functional_jax_env import FunctionalJaxEnv
+from gymnasium.utils import EzPickle, seeding
+from gymnasium.wrappers import HumanRendering
+
+
+RenderStateType = Tuple["pygame.Surface", str, int]  # type: ignore  # noqa: F821
+
+
+deck = jnp.array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 10, 10, 10])
+
+
+def obs_from_device(obs):
+    """Converts a tuple of device arrays into a tutple of ints."""
+    return tuple([int(jax.device_get(obs[i])) for i in range(len(obs))])
+
+
+def cmp(a, b):
+    """Returns 1 if a > b, otherwise returns -1."""
+    return (a > b).astype(int) - (a < b).astype(int)
+
+
+def random_card(key):
+    """Draws a randowm card (with replacement)."""
+    key = random.split(key)[0]
+    choice = random.choice(key, deck, shape=(1,))
+
+    return choice[0].astype(int), key
+
+
+def draw_hand(key, hand):
+    """Draws a starting hand of two random cards."""
+    new_card, key = random_card(key)
+    hand = hand.at[0].set(new_card)
+    new_card, key = random_card(key)
+    hand = hand.at[1].set(new_card)
+    return hand, key
+
+
+def draw_card(key, hand, index):
+    """Draws a new card and adds it to a hand."""
+    new_card, key = random_card(key)
+    hand = hand.at[index].set(new_card)
+    return key, hand, index + 1
+
+
+def usable_ace(hand):
+    """Checks to se if a hand has a usable ace."""
+    return jnp.logical_and((jnp.count_nonzero(hand == 1) > 0), (sum(hand) + 10 <= 21))
+
+
+def take(env_state):
+    """This function is called if the player has decided to take a card."""
+    state, key = env_state
+    dealer_hand = state[0]
+    player_hand = state[1]
+    dealer_cards = state[2]
+    player_cards = state[3]
+    key, new_player_hand, _ = draw_card(key, player_hand, player_cards)
+
+    return (dealer_hand, new_player_hand, dealer_cards, player_cards + 1), key
+
+
+def dealer_stop(val):
+    """This function determines if the dealer should stop drawing."""
+    return sum_hand(val[1]) < 17
+
+
+def draw_card_wrapper(val):
```

Comment:
Why is this needed?

Reply:
it's a hack to allow the draw_card function to accept a tuple of it's arg since the while loop eval structure only allows a single object to get passed into the while loop body function, and the draw_card function is used elsewhere with multiple args, so it seem unreasonable to make draw_card take only a single tuple as an argument. https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.while_loop.html